cmd/utils/executor.go

```diff
@@ -0,0 +1,247 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package utils
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/manifoldco/promptui"
+	"github.com/manifoldco/promptui/screenbuf"
+	"github.com/okteto/okteto/pkg/log"
+	"golang.org/x/term"
+)
+
+type ManifestExecutor interface {
+	Execute(command string, env []string) error
+}
+
+type Executor struct {
+	outputMode string
+	displayer  executorDisplayer
+}
+
+type executorDisplayer interface {
+	display(scanner *bufio.Scanner, command string, sb *screenbuf.ScreenBuf)
+	startCommand(cmd *exec.Cmd) (io.Reader, error)
+}
+
+type ttyExecutorDisplayer struct{}
+type plainExecutorDisplayer struct{}
+type jsonExecutorDisplayer struct{}
+
+type jsonMessage struct {
+	Level     string `json:"level"`
+	Stage     string `json:"stage"`
+	Message   string `json:"message"`
+	Timestamp int64  `json:"timestamp"`
+}
+
+// NewExecutor returns a new executor
+func NewExecutor(output string) *Executor {
+	if output == "tty" && !isSupportForTTY() {
+		output = "plain"
+	}
+	var displayer executorDisplayer
+	switch output {
+	case "tty":
+		displayer = ttyExecutorDisplayer{}
+	case "plain":
+		displayer = plainExecutorDisplayer{}
+	case "json":
+		displayer = jsonExecutorDisplayer{}
+	}
+	return &Executor{
+		outputMode: output,
+		displayer:  displayer,
+	}
+}
+
+// Execute executes the specified command adding `env` to the execution environment
+func (e *Executor) Execute(command string, env []string) error {
+
+	cmd := exec.Command("bash", "-c", command)
+	cmd.Env = append(os.Environ(), env...)
+
+	reader, err := e.displayer.startCommand(cmd)
+	if err != nil {
+		return err
+	}
+
+	scanner := bufio.NewScanner(reader)
+
+	sb := screenbuf.New(os.Stdout)
+	go e.displayer.display(scanner, command, sb)
+
+	err = cmd.Wait()
+	if e.outputMode == "tty" {
+		collapseTTY(command, err, sb)
+	}
+	return err
+}
+
+func (plainExecutorDisplayer) startCommand(cmd *exec.Cmd) (io.Reader, error) {
+	reader, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := cmd.Start(); err != nil {
+		return nil, err
+	}
+	return reader, nil
+}
+
+func (plainExecutorDisplayer) display(scanner *bufio.Scanner, _ string, _ *screenbuf.ScreenBuf) {
+	for scanner.Scan() {
+		line := scanner.Text()
+		fmt.Println(line)
+	}
+}
+
+func (jsonExecutorDisplayer) startCommand(cmd *exec.Cmd) (io.Reader, error) {
+	reader, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := cmd.Start(); err != nil {
+		return nil, err
+	}
+	return reader, nil
```

Comment:
This logic can be extracted to a common function to be shared by `jsonExecutorDisplayer`, `plainExecutorDisplayer` and `ttyExecutorDisplayer` for windows